### ISSUE TITLE
Constructor argument for `edm::GlobalCache<T>`

### ISSUE THREAD

**Issue by schneiml** [cms-sw/cmssw#27125]
While trying to convert `DQMEDAnalyzer` to `edm::stream`, I hit the following problem:
- To share state between the per-stream instances of `edm::stream` modules, I need to use a `edm::GlobalCache<T>`.
- `edm::GlobalCache<T>` changes the _constructor_ signature used by the EDM framework. [0]
- This means now ~400 derived CMSSW plugins would need to be changed to accept the additional argument.

To make things worse
- the type of the new argument is deeply internal to the DQM framework and should not be exposed to the outside, yet it would need to be mentioned in each of the subsystem modules.
- the argument passed is rather useless, since `globalCache()` gives access to the same value.
 
Would it be possible to make passing this argument optional? Alternatively, I'd have to find a way to share the required information without a `GlobalCache`, which is probably not a good idea either.

[0] https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkStreamModuleInterface#edm_GlobalCacheT

**Comment by Dr15Jones** [comment]
assign core

**Comment by Dr15Jones** [comment]
> the argument passed is rather useless, since globalCache() gives access to the same value.

That is incorrect. If you call `globalCache()` in the constructor, you get back nullptr.  The value is not set until after the constructor finishes (which hopefully makes sense once you think about it).

It is possible to override the framework behavior for a particular `GlobalCache` type. To do so add the following to your `DQMEDAnalyzer.h` file before the declaration of the `EDAnalyzer` but after at least forward declaring your particular `GlobalCache` type (which I've denoted below using the type name `TheDQMGlobalCache`)

```C++
#incude "FWCore/Framework/interface/stream/makeGlobal.h"

namespace edm::stream::impl {
template< typename T>
T* makeStreamModule(edm::ParameterSet const& iPSet, TheDQMGlobalCache const*) {
  return new T(iPSet); 
}
}
```

**Comment by schneiml** [comment]
> If you call globalCache() in the constructor, you get back nullptr.

Ah, I did not expect that, given that the `globalCache` is supposedly initialized before the constructor is called. Anyways I don't really care, since I only need the cache around `beginStream`. [1]

The suggestion of specializing the template probably works, though I would not use such intrusive code without approval of cmssw-core. I'll see how far that snippet gets us. Thanks!

[1] That is at least what I have implemented so far. It would be nice to initialize things earlier, but I'd need at least the number of modules to be created and a unique id for each (read: number of streams and streamId), which I suspect are not available at the time of the constructor calls.
